fda/metrics.py

```diff
@@ -0,0 +1,502 @@
+
+import scipy.integrate
+import numpy
+
+from . import FDataGrid
+from .registration import (normalize_warping, _normalize_scale,
+                           to_srsf, elastic_registration_warping)
+
+
+
+def norm_lp(fdatagrid, p=2):
+    r"""Calculate the norm of all the samples in a FDataGrid object.
+
+    For each sample sample f the lp norm is defined as:
+
+    .. math::
+        \lVert f \rVert = \left( \int_D \lvert f \rvert^p dx \right)^{
+        \frac{1}{p}}
+
+    Where D is the domain over which the functions are defined.
+
+    The integral is approximated using Simpson's rule.
+
+    Args:
+        fdatagrid (FDataGrid): FDataGrid object.
+        p (int, optional): p of the lp norm. Must be greater or equal
+            than 1. Defaults to 2.
+
+    Returns:
+        numpy.darray: Matrix with as many rows as samples in the first
+        object and as many columns as samples in the second one. Each
+        element (i, j) of the matrix is the inner product of the ith sample
+        of the first object and the jth sample of the second one.
+
+    Examples:
+        Calculates the norm of a FDataGrid containing the functions y = 1
+        and y = x defined in the interval [0,1].
+
+
+        >>> x = numpy.linspace(0,1,1001)
+        >>> fd = FDataGrid([numpy.ones(len(x)), x] ,x)
+        >>> norm_lp(fd).round(2)
+        array([ 1.  ,  0.58])
+
+        The lp norm is only defined if p >= 1.
+
+        >>> norm_lp(fd, p = 0.5)
+        Traceback (most recent call last):
+            ....
+        ValueError: p must be equal or greater than 1.
+
+    """
+    # Checks that the lp normed is well defined
+    if p < 1:
+        raise ValueError("p must be equal or greater than 1.")
+
+    if fdatagrid.ndim_image > 1:
+        raise ValueError("Not implemented for image with "
+                         "dimension greater than 1")
+
+    # Computes the norm, approximating the integral with Simpson's rule.
+    return scipy.integrate.simps(numpy.abs(fdatagrid.data_matrix[..., 0]) ** p,
+                                 x=fdatagrid.sample_points
+                                 ) ** (1 / p)
+
+
+def metric(fdatagrid, fdatagrid2, norm=norm_lp, **kwargs):
```

Comment:
Not all metrics come from a norm. I know that this function is from Miguel, but maybe we should change its name. In fact, I would split it into two functions:
- A small function that returns a distance given a norm, such as:
```python
def distance_from_norm(norm, **kwargs):
    def norm_distance(fdata1, fdata2):
        return norm(fdata1 - fdata2, **kwargs)
    
    norm_distance.__name__ = f"{norm.__name__}_distance"
    return norm_distance
```
- A function `pairwise_distances` that receives a distance (not a norm) and computes the pairwise distances.

Then each distance function should only need to care about distance between individual functions, and the pairwise distances  could be delegated to this one.

Reply:
Changes added in 0fc7096. Now the distances will receive two FDatas with one sample and returns a float with the distance. Test for the metrics were added too.